packages/next-auth/src/lib/index.ts

```diff
@@ -0,0 +1,239 @@
+  NextApiRequest,
+  NextApiResponse,
+} from "next"
+import type { AppRouteHandlerFn } from "next/dist/server/future/route-modules/app-route/module"
+import type { NextFetchEvent, NextMiddleware, NextRequest } from "next/server"
+
+/** Configure NextAuth.js. */
+export interface NextAuthConfig extends AuthConfig {
+  /**
+   * Callbacks are asynchronous functions you can use to control what happens when an auth-related action is performed.
+   * Callbacks **allow you to implement access controls without a database** or to **integrate with external databases or APIs**.
+   */
+  callbacks?: AuthConfig["callbacks"] & {
+    /**
+     * Invoked when a user needs authorization, using [Middleware](https://nextjs.org/docs/advanced-features/middleware).
+     *
+     * You can override this behavior by returning a {@link NextResponse}.
+     *
+     * @example
+     * ```ts title="app/auth.ts"
+     * ...
+     * async authorized({ request, auth }) {
+     *   const url = request.nextUrl
+     *
+     *   if(request.method === "POST") {
+     *     const { authToken } = (await request.json()) ?? {}
+     *     // If the request has a valid auth token, it is authorized
+     *     const valid = await validateAuthToken(authToken)
+     *     if(valid) return true
+     *     return NextResponse.json("Invalid auth token", { status: 401 })
+     *   }
+     *
+     *   // Logged in users are authenticated, otherwise redirect to login page
+     *   return !!auth.user
+     * }
+     * ...
+     * ```
+     *
+     * :::warning
+     * If you are returning a redirect response, make sure that the page you are redirecting to is not protected by this callback,
+     * otherwise you could end up in an infinite redirect loop.
+     * :::
+     */
+    authorized?: (params: {
+      /** The request to be authorized. */
+      request: NextRequest
+      /** The authenticated user or token, if any. */
+      auth: Session | null
+    }) => Awaitable<boolean | NextResponse | undefined>
+  }
+}
+
+function detectOrigin(_headers: Headers | ReturnType<typeof headers>) {
+  const host = _headers.get("x-forwarded-host") ?? _headers.get("host")
+  const protocol =
+    _headers.get("x-forwarded-proto") === "http" ? "http" : "https"
+  return new URL(process.env.NEXTAUTH_URL ?? `${protocol}://${host}`)
+}
+
+/** Server-side method to read the session. */
+async function getSession(headers: Headers, config: NextAuthConfig) {
+  const origin = detectOrigin(headers)
+  const request = new Request(`${origin}/api/auth/session`, {
+    headers: { cookie: headers.get("cookie") ?? "" },
+  })
+  config.useSecureCookies ??= origin.protocol === "https:"
+
+  // Since we are server-side, we don't need to filter out the session data
+  // See https://nextjs.authjs.dev/v5#authenticating-server-side
+  config.callbacks = Object.assign(config.callbacks ?? {}, {
+    session({ session, user, token }) {
+      return { ...session, user: user ?? token }
+    },
```

Comment:
@balazsorban44 Why are you overwriting the user-defined session callback here? This will also overwrite the `config` used by the http-handlers meaning that if you call `[GET] /api/auth/session` that you will expose the full token object, so also the sensitive information.

Reply:
See the comment right above it:

https://github.com/nextauthjs/next-auth/blob/790f7617083d055b94d05c5c815749258da84852/packages/next-auth/src/lib/index.ts#L75-L76

This method (`getSession`) is an internal function only invoked where we are server-side.

`/api/auth/session` behavior does not change.